File: April 2025/07.py
class Solution:
    def isCycle(self, V, edges):
        # Create adjacency list representation of the graph
        adj = [[] for _ in range(V)]
        for u, v in edges:
            adj[u].append(v)
        
        # Initialize visited and recursion stack
        visited = [False] * V
        recStack = [False] * V
        
        # Helper function to perform DFS
        def dfs(node):
            visited[node] = True
            recStack[node] = True
            
            # Traverse neighbors
            for neighbor in adj[node]:
                if not visited[neighbor]:
                    if dfs(neighbor):  # Recursive DFS call
                        return True
                elif recStack[neighbor]:  # Cycle detected
                    return True
            
            recStack[node] = False  # Backtrack
            return False
        
        # Check for cycles in each component
        for i in range(V):
            if not visited[i]:
                if dfs(i):
                    return True
        
        return False

File: April 2025/07.cpp
class Solution {
  public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // Create adjacency list representation of the graph
        vector<vector<int>> adj(V);
        for (auto edge : edges) {
            adj[edge[0]].push_back(edge[1]);
        }

        // Initialize visited and recursion stack
        vector<bool> visited(V, false);
        vector<bool> recStack(V, false);

        // Helper function to perform DFS
        function<bool(int)> dfs = [&](int node) {
            visited[node] = true;
            recStack[node] = true;

            // Traverse neighbors
            for (int neighbor : adj[node]) {
                if (!visited[neighbor]) {
                    if (dfs(neighbor)) { // Recursive DFS call
                        return true;
                    }
                } else if (recStack[neighbor]) { // Cycle detected
                    return true;
                }
            }

            recStack[node] = false; // Backtrack
            return false;
        };

        // Check for cycles in each component
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                if (dfs(i)) {
                    return true;
                }
            }
        }
        return false;
    }
};
